### ISSUE TITLE
Interest in a namespace declaration sniff to check scoped/no name usage ?

### ISSUE THREAD

**Issue by jrfnl** [squizlabs/PHP_CodeSniffer#2324]
Is there interest in a namespace declaration sniff which would check for things like:

* Scoped namespaces _and forbid their use_.
* Namespace declarations without a namespace name (= global namespace), i.e `namespace;`, _and forbid their use_.
* Multiple (scoped) namespaces in a file _and forbid this_.

I wrote a sniff for this for an in-company standard already and if there is interest, they're happy to let me pull it here.

**Comment by kwhat** [comment]
Hi @jrfnl,

First, what is a "scoped namespace?"  I cannot find this verbiage in the PHP docs.  

As for [multiple namespaces](https://www.php.net/manual/en/language.namespaces.definitionmultiple.php#language.namespaces.definitionmultiple) in the same file, I am going to argue against adding a filter for it.  Let me explain:

The only way to mock built-in PHP functions for unit testing is using these Multiple (bracketed is the word you were looking for) namespaces.  For example, I recently needed to mock `curl_multi_exec()` so that I could have reliable unit tests for some async request code.  The only way to create a unit test class to do this is outlined below.  I am all for the global namespace restriction, and restrictions on how multiple namespaces are used, like only with brackets.

On a related note, the reason I came across this bug was that the PSR-2 rules have side effects with the aforementioned code.  Currently, the use declarations in these bracketed namespaces produces a "USE declarations must go after the first namespace declaration" which is clearly not going to work in this case.

```
<?php

namespace My\Project\Http {

    use My\Project\Test\Http\ClientTest;
    use ReflectionException;
    use ReflectionObject;

    /**
     * @param resource $mh
     * @param int $still_running
     *
     * @return int
     * @throws ReflectionException
     */
    function curl_multi_exec($mh, &$still_running)
    {
        $still_running = 1;

        $ref = new ReflectionObject(ClientTest::$client);
        $method = $ref->getMethod("writeHeader");
        $method->setAccessible(true);
        $method->invoke(ClientTest::$client, ClientTest::$handle, "HTTP/1.1 200 OK\r\n");
        $method->invoke(ClientTest::$client, ClientTest::$handle, "Content-Type: application/json\r\n");

        $method = $ref->getMethod("writeBody");
        $method->setAccessible(true);
        $method->invoke(ClientTest::$client, ClientTest::$handle, "{}");

        return CURLM_OK;
    }

    /**
     * @param resource $mh
     * @param int|null $msgs_in_queue
     *
     * @return array
     */
    function curl_multi_info_read($mh, &$msgs_in_queue = null)
    {
        return array(
            "msg" => CURLMSG_DONE,
            "result" => ClientTest::$result,
            "handle" => ClientTest::$handle
        );
    }
}

namespace My\Project\Test\Http {

    use My\Project\Futures\PromiseInterface;
    use My\Project\Http\Client;
    use My\Project\Http\Exceptions\MethodNotAllowedException;
    use My\Project\Http\Request;
    use Exception;
    use PHPUnit\Framework\TestCase;
    use ReflectionException;
    use ReflectionObject;
    use Slim\Http\Headers;
    use Slim\Http\RequestBody;
    use Slim\Http\Uri;

    class ClientTest extends TestCase
    {
        /** @var Client $client */
        public static $client;

        /** @var int $result */
        public static $result = CURLE_OK;

        /** @var resource|null $handle */
        public static $handle = null;

        /**
         * @throws MethodNotAllowedException
         * @throws ReflectionException
         * @throws Exception
         */
        public function testSendSuccess()
        {
            $request = new Request(
                "GET",
                Uri::createFromString("/"),
                new Headers(),
                array(),
                array("SERVER_PROTOCOL" => "1.1"),
                new RequestBody()
            );

            static::$client = new Client();
            $promise = static::$client->send($request);
            $this->assertInstanceOf(PromiseInterface::class, $promise);

            $ref = new ReflectionObject(static::$client);
            $property = $ref->getProperty("promises");
            $property->setAccessible(true);
            $promises = $property->getValue(static::$client);

            $key = array_search($promise, $promises, true);
            $this->assertNotFalse($key);

            $property = $ref->getProperty("handles");
            $property->setAccessible(true);
            $handles = $property->getValue(static::$client);

            static::$handle = $handles[$key];

            $promise->wait();
        }

        public function testGetConfigSuccess()
        {
            $client = new Client();
            $this->assertIsArray($client->getConfig());
            $this->assertIsInt($client->getConfig("cache_size"));
            $this->assertNull($client->getConfig("invalid"));
        }
    }
}
```

**Comment by jrfnl** [comment]
> First, what is a "scoped namespace?" I cannot find this verbiage in the PHP docs.

Curly brackets - in PHPCS - create _scope conditions_ for the code contained within it. That's what the term is about.


> As for multiple namespaces in the same file, I am going to argue against adding a filter for it. Let me explain:

Sorry, but your arguments are irrelevant to this issue.
1. This issue is not about whether you should **_use_** the sniff, only about whether there is **_interest_** in such a sniff.
2. The sniff would go into `Generic` and would not be added to any of the PHPCS native standards, so by default, it wouldn't be used, _unless you specifically_ **choose** _to turn in on_.

In other words, once it would be added to PHPCS, it would be possible for people to _choose_ to add it to their standard, but if you don't like, it's simple: just don't add it.

**Comment by kwhat** [comment]
You are correct.

**Comment by jrfnl** [comment]
I'm going to close this issue. While these sniffs ended up not being pulled to PHPCS itself, they did get written and are now publicly available via the [PHPCSExtra](https://github.com/PHPCSStandards/PHPCSExtra/) package (Packagist/Composer: `phpcsstandards/phpcsextra`).

The sniffs are called:
* `Universal.Namespaces.DisallowDeclarationWithoutName`
* `Universal.Namespaces.DisallowCurlyBraceSyntax`
* `Universal.Namespaces.EnforceCurlyBraceSyntax`
* `Universal.Namespaces.OneDeclarationPerFile`
